### PR TITLE
support for marigold

### PR DESCRIPTION
**Pull request by affromero**
Discussion: https://github.com/thygate/stable-diffusion-webui-depthmap-script/discussions/379
Issue: https://github.com/thygate/stable-diffusion-webui-depthmap-script/issues/383
I continued the implementation and this is an example:
|                 Input                      | Leres No Boost | Marigold No Boost |
|:-------------------------------------:|:-------------------------------------:|:-------------------------------------:|
| |   |   |
|                                      | Leres Boosted | Marigold Boosted |
|                                      |  |  |

### CODE REVIEW

**Comment by affromero** [comment]
|                 Input                      | Leres Boosted | Marigold No Boost | Marigold boosted |
|:-------------------------------------:|:-------------------------------------:|:-------------------------------------:|:-------------------------------------:|
| | | | |

File: src/depthmap_generation.py [hunk 1]
```diff
@@ -418,13 +418,19 @@ def estimatemarigold(image, model, w, h, device):
     tol = 1e-3
     reduction_method = "median"
     merging_max_res = None
+    resize_to_max_res = None
 
     # From Marigold repository run.py
     with torch.no_grad():
+        image = (image * 255).astype(np.uint8)
```
Comment (graemeniedermayer) [inline src/depthmap_generation.py:425]: We should be able to simplify by removing
` image = (image * 255).astype(np.uint8)`
 and turning line 431
```
rgb = np.transpose(image, (2, 0, 1))  # [H, W, rgb] -> [rgb, H, W]
rgb_norm = rgb / 255.0
```
into
`rgb_norm = np.transpose(image, (2, 0, 1))  # [H, W, rgb] -> [rgb, H, W]`

That way we aren't multiplying and dividing 255.

Comment (affromero) [reply src/depthmap_generation.py:425]: The rationale behind that is that in Marigold they [resize](https://github.com/prs-eth/Marigold/blob/main/run.py#L140) the image, and the input of that function is `PIL.Image`, so I just thought of having the same option. But yeah, maybe then multiplying and dividing in L427?

Comment (graemeniedermayer) [reply src/depthmap_generation.py:425]: It does look cleaner to resize with a numpy array using something like
`image = cv2.resize(image, (h, w), interpolation=cv2.INTER_CUBIC)`.

Although I'm not sure if there's any subtle differences between cv2 resize and PIL.Image resize.

Great changes overall!

**Comment by semjon00** [comment]
Hello and thank you so much for this! I will test as soon as I have time. Meanwhile I will ask for some clarifications...

File: pix2pix/models/pix2pix4depth_model.py [hunk 2]
```diff
@@ -94,8 +94,8 @@ def set_input_train(self, input):
         self.real_A = torch.cat((self.outer, self.inner), 1)
 
     def set_input(self, outer, inner):
-        inner = torch.from_numpy(inner).unsqueeze(0).unsqueeze(0)
-        outer = torch.from_numpy(outer).unsqueeze(0).unsqueeze(0)
+        inner = torch.from_numpy(inner).unsqueeze(0).unsqueeze(0).float()
```
Comment (semjon00) [inline pix2pix/models/pix2pix4depth_model.py:97]: Are you super-duper sure this does not break anything?

File: src/depthmap_generation.py (hunk 1 above)
Comment (semjon00) [reply src/depthmap_generation.py:425]: I have no strong feeling either way.

**Comment by semjon00** [comment]
Overall looks good. I almost did not believe that I stopped short of only 16 lines with my branch - let alone adding boost in the same 16 lines.

**Comment by semjon00** [comment]
My only concern is that the depthmaps are inverted. In this script, white=near convention is used. I think the only thing to fix that would be adding Marigold (10) to line 298 in depthmap_generation.py: `raw_prediction_invert = self.depth_model_type in [0, 7, 8, 9]`. However, after changing, please test if it would actually work nicely, also that it will it generate sensible stereoimages, etc.

**Comment by semjon00** [comment]
Also, would you be so kind to append README with Marigold mentions and an Acknowledgement? Before pushing Marigold support to main we will also need a version bump (misc.py) and a changelog ([CHANGELOG.md](https://github.com/thygate/stable-diffusion-webui-depthmap-script/blob/main/CHANGELOG.md)) with one change.

**Comment by graemeniedermayer** [comment]
> My only concern is that the depthmaps are inverted. In this script, white=near convention is used. I think the only thing to fix that would be adding Marigold (10) to line 298 in depthmap_generation.py: `raw_prediction_invert = self.depth_model_type in [0, 7, 8, 9]`. However, after changing, please test if it would actually work nicely, also that it will it generate sensible stereoimages, etc.

I have tested this. It works without boost.

Maybe Marigold should be git clone into "extensions/stable-diffusion-webui-depthmap-script/Marigold" rather than "https://github.com/prs-eth/Marigold" to avoid future conflicts (this is why midas is installed there). Also I believe `diffusers` is now a requirement.

**Comment by semjon00** [comment]
@graemeniedermayer Indeed, it needs `diffusers>=0.20.1`, should be added to `install.py`. Thankfully, it seems to be a rather lightweight dependency. About conflits - I'm not sure about it... maybe?

**Comment by graemeniedermayer** [comment]
> @graemeniedermayer Indeed, it needs `diffusers>=0.20.1`, should be added to `install.py`. Thankfully, it seems to be a rather lightweight dependency. About conflits - I'm not sure about it... maybe?

Also marigold standalone mode doesn't function with the current imports.

**Comment by semjon00** [comment]
Oh, supporting standalone is definitely needed.

**Comment by affromero** [comment]
Hey folks, I added some of your suggested changes. I am not entirely sure if there was a protocol for the requirements/install files, so let me know what should I change. In particular, I use this project as standalone, so the `install.py` I am not sure about it.

**Comment by graemeniedermayer** [comment]
I think this should be merged onto the marigold branch. There might be some small changes/extra testing to do before merging into main.

Great work!

**Comment by semjon00** [comment]
Merged and did some fixes. Now it would be awesome to support automatic repository pull for standalone. Yes, we can just copy all the files into the root again... But I'd rather not have any new code linked this way.

**Comment by aulerius** [comment]
This is really good news!

I also want do draw some attention to equivalent implementation in ComfyUI, and that it implements floating point EXR export:

> I added a remap node to see the full range better, and OpenEXR node to save the full range, works wonders compared to default png when used in VFX/3D modeling software.

Which relates to #372 and #370
Could it be a nice opportunity to try that as well?

**Comment by semjon00** [comment]
Not now, I am reeeaaally busy.

**Comment by semjon00** [comment]
@affromero @graemeniedermayer I can't get it to work! Please help... :(
I get things like
`WARNING:xformers:WARNING[XFORMERS]: xFormers can't load C++/CUDA extensions. xFormers was built for:
    PyTorch 2.1.1+cu121 with CUDA 1201 (you have 2.1.1+cpu)
    Python  3.10.11 (you have 3.10.6)
  Please reinstall xformers (see https://github.com/facebookresearch/xformers#installing-xformers)
  Memory-efficient attention, SwiGLU, sparse and more won't be available.
  Set XFORMERS_MORE_DETAILS=1 for more details
`
For whatever reason trying to have XFORMERS automatically installs the wrong torch version

GPU: 1080Ti
OS: Windows 10

**Comment by graemeniedermayer** [comment]
Oh I just realized I tested it with a different `install.py` file maybe removing all the new requirements besides `diffusers` would be best to avoid conflicts. The others are very likely to cause conflicts with other repos.

**Comment by graemeniedermayer** [comment]
> I also want do draw some attention to equivalent implementation in ComfyUI, and that it implements floating point EXR export:

It shouldn't be too challenging to save the numpy arrays is there a library for converting to EXR?

**Comment by aulerius** [comment]
> > I also want do draw some attention to equivalent implementation in ComfyUI, and that it implements floating point EXR export:
> 
> It shouldn't be too challenging to save the numpy arrays is there a library for converting to EXR?

Apparently so, simply called "OpenEXR".
[This is how it's done](https://github.com/kijai/ComfyUI-Marigold/blob/83b347a52bb8b6bc3486fb2db5d5472917e0e27d/nodes.py#L206C24-L206C24) in the aforementioned ComfyUI implementation.
And TIF files also support floating point 32bit precision.

**Pull request merged**